File: reverse_ips.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"strings"
	"sync"
)

type ReverseIPLookup struct {
	domainSet map[string]struct{}
}

func NewReverseIPLookup() *ReverseIPLookup {
	return &ReverseIPLookup{
		domainSet: make(map[string]struct{}),
	}
}

func (r *ReverseIPLookup) fetchDomainsFromHackerTarget(ip string) ([]string, error) {
	url := fmt.Sprintf("https://api.hackertarget.com/reverseiplookup/?q=%s", ip)
	resp, err := http.Get(url)
	if err != nil {
		return nil, fmt.Errorf("HackerTarget API error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HackerTarget API non-OK status: %d", resp.StatusCode)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("error reading HackerTarget response: %w", err)
	}

	domains := strings.Split(strings.TrimSpace(string(body)), "\n")
	return domains, nil
}

func (r *ReverseIPLookup) fetchDomainsFromSecurityTrails(ip, apiKey string) ([]string, error) {
	url := fmt.Sprintf("https://api.securitytrails.com/v1/ips/%s/domains", ip)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("SecurityTrails request creation error: %w", err)
	}
	req.Header.Add("apikey", apiKey)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("SecurityTrails API error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("SecurityTrails API non-OK status: %d", resp.StatusCode)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("error reading SecurityTrails response: %w", err)
	}

	var data struct {
		Domains []string `json:"domains"`
	}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, fmt.Errorf("error parsing SecurityTrails JSON: %w", err)
	}

	return data.Domains, nil
}

func (r *ReverseIPLookup) processSingleIP(ip, apiKey string, wg *sync.WaitGroup) {
	defer wg.Done()

	domainSet := make(map[string]struct{})

	hackerTargetDomains, err := r.fetchDomainsFromHackerTarget(ip)
	if err != nil {
		log.Printf("HackerTarget error for IP %s: %v", ip, err)
	} else {
		for _, domain := range hackerTargetDomains {
			domainSet[domain] = struct{}{}
		}
	}

	if apiKey != "" {
		securityTrailsDomains, err := r.fetchDomainsFromSecurityTrails(ip, apiKey)
		if err != nil {
			log.Printf("SecurityTrails error for IP %s: %v", ip, err)
		} else {
			for _, domain := range securityTrailsDomains {
				domainSet[domain] = struct{}{}
			}
		}
	}

	for domain := range domainSet {
		r.domainSet[domain] = struct{}{}
	}
}

func (r *ReverseIPLookup) processAllIPs(ips []string, apiKey string) {
	var wg sync.WaitGroup

	for _, ip := range ips {
		if strings.TrimSpace(ip) == "" {
			continue
		}
		wg.Add(1)
		go r.processSingleIP(ip, apiKey, &wg)
	}

	wg.Wait()
}

func (r *ReverseIPLookup) saveToFile(filename string) {
	file, err := ioutil.TempFile(".", filename)
	if err != nil {
		log.Fatalf("Could not create file: %v", err)
	}
	defer file.Close()

	for domain := range r.domainSet {
		if _, err := file.WriteString(fmt.Sprintf("%s\n", domain)); err != nil {
			log.Fatalf("Could not write to file: %v", err)
		}
	}
	fmt.Printf("Results saved to %s\n", filename)
}

func main() {
	inputFile := "iptables.txt"
	outputFile := "domains_found.txt"
	apiKey := "YOUR_API_KEY"

	data, err := ioutil.ReadFile(inputFile)
	if err != nil {
		log.Fatalf("Error reading input file: %v", err)
	}
	ips := strings.Split(strings.TrimSpace(string(data)), "\n")

	lookup := NewReverseIPLookup()
	lookup.processAllIPs(ips, apiKey)
	lookup.saveToFile(outputFile)
}

File: reverse_ips.py
import asyncio
import aiohttp
from typing import List, Set


class ReverseIPLookup:
    """Reverse IP lookup using multiple services"""

    def __init__(self) -> None:
        """Initialize the ReverseIPLookup object"""
        self.unique_domains: Set[str] = set()

    async def get_domains_hackertarget(self, ip: str, session: aiohttp.ClientSession) -> List[str]:
        """Using HackerTarget API to get reverse IP data

        Args:
            ip (str): IP address to look up
            session (aiohttp.ClientSession): Session to use for the request

        Returns:
            List[str]: List of domains found for the given IP
        """
        url = f"https://api.hackertarget.com/reverseiplookup/?q={ip}"
        try:
            async with session.get(url) as response:
                if response.status == 200:
                    text = await response.text()
                    return [domain.strip() for domain in text.split('\n') if domain.strip()]
                return []
        except Exception as e:
            print(f"Error with HackerTarget API: {e}")
            return []

    async def get_domains_securitytrails(self, ip: str, session: aiohttp.ClientSession, api_key: str) -> List[str]:
        """Using SecurityTrails API to get reverse IP data

        Args:
            ip (str): IP address to look up
            session (aiohttp.ClientSession): Session to use for the request
            api_key (str): SecurityTrails API key

        Returns:
            List[str]: List of domains found for the given IP
        """
        url = f"https://api.securitytrails.com/v1/ips/{ip}/domains"
        headers = {"apikey": api_key}
        try:
            async with session.get(url, headers=headers) as response:
                if response.status == 200:
                    data = await response.json()
                    return data.get('domains', [])
                return []
        except Exception as e:
            print(f"Error with SecurityTrails API: {e}")
            return []

    async def process_ip(self, ip: str, session: aiohttp.ClientSession, security_trails_api_key: str = None) -> Set[str]:
        """Process single IP address using multiple services

        Args:
            ip (str): IP address to look up
            session (aiohttp.ClientSession): Session to use for the request
            security_trails_api_key (str, optional): SecurityTrails API key. Defaults to None.

        Returns:
            Set[str]: Set of unique domains found for the given IP
        """
        print(f"Processing IP: {ip}")
        domains = set()

        # Get domains from HackerTarget
        ht_domains = await self.get_domains_hackertarget(ip, session)
        domains.update(ht_domains)

        # If SecurityTrails API key is provided, use it as well
        if security_trails_api_key:
            st_domains = await self.get_domains_securitytrails(ip, session, security_trails_api_key)
            domains.update(st_domains)

        self.unique_domains.update(domains)
        return domains

    async def process_ips(self, ips: List[str], security_trails_api_key: str = None) -> None:
        """Process multiple IP addresses concurrently

        Args:
            ips (List[str]): List of IP addresses to look up
            security_trails_api_key (str, optional): SecurityTrails API key. Defaults to None.
        """
        async with aiohttp.ClientSession() as session:
            tasks = []
            for ip in ips:
                # Add delay between requests to avoid rate limiting
                await asyncio.sleep(1)
                tasks.append(self.process_ip(ip, session, security_trails_api_key))
            await asyncio.gather(*tasks)

    def save_results(self, output_file: str) -> None:
        """Save results to file

        Args:
            output_file (str): File to save the results to
        """
        with open(output_file, 'w') as f:
            for domain in sorted(self.unique_domains):
                f.write(f"{domain}\n")
        print(f"Results saved to {output_file}")

async def main() -> None:
    """Main function to run the script"""
    input_file = "iptables.txt"
    output_file = "domains_found.txt"
    security_trails_api_key = "YOUR_API_KEY"

    with open(input_file, 'r') as f:
        ips = [line.strip() for line in f if line.strip()]

    lookup = ReverseIPLookup()
    await lookup.process_ips(ips, security_trails_api_key)
    lookup.save_results(output_file)

if __name__ == "__main__":
    asyncio.run(main())
